### DISCUSSION TITLE
guide/international/

### DISCUSSION THREAD

**Discussion by giscus** [pengzhanbo/vuepress-theme-plume#234] [docs-comment]
# guide/international/

目录 要使用内置的 i18n (国际化) 功能，需要创建类似于下面的目录结构： 重要 在不同的语言目录下，请尽量保持文件名和目录名的一致。这有助于在 切换语言时，主题能够正确的导航到 文章 的不同的语言版本。 vuepress 配置 在 .vuepress/config.js 中配置： 主题配置 在本主题中，同样使用 locales 配置项进行多语言配...

https://theme-plume.vuejs.press/guide/international/

**Comment by MrHaoH** [comment]
大佬，请教个问题，我如果想配置自定义组件中的多语言的话，之前都是用el-plus中的 el-config-provider 组件，“
<el-config-provider :locale="zhCn">
    <app />
  </el-config-provider>

**Reply by MrHaoH** [reply]
用主题的话，怎么实现？感谢大佬的回复

**Reply by pengzhanbo** [reply]
这可能会比较麻烦，vuePress 和 主题都没有使用如 `vue-i18n` 之类的包。主题内部是先定义一个 `const locales = { '/': {}, '/zh/': {} }` 的对象保存多语言数据，然后通过 `vuepress/client` 导出的 `useRouteLocale` 获取当前页面的多语言根链接，比如 `/` 或 `/zh/` 等，再从 `locales` 上获取对应的数据。

**Reply by MrHaoH** [reply]
哦哦，好的，感谢大佬的回复

**Comment by MrHaoH** [comment]
大佬，我怎么添加一种语言呢，好比繁体中文，我在.vuepress/config.ts文件中，defineUserConfig.locales中添加了
'/zh-TW/': {
            title: '',
            lang: 'zh-TW',
            description: '',
            ...zh_tw
        }
但是报错了，这是为啥

**Reply by MrHaoH** [reply]
哦哦，我好像知道了，必须得在zh-TW文件夹下面创建ReadME文件是吧，我去试试

**Comment by MrHaoH** [comment]
大佬，我找了半天了，没找出来，跪求大佬帮助
为什么我生成的docs\.vuepress\.temp\internal\themePlumeData.js文件中的export const themeData = {
    "locales": {
        "/tw/": { 
            "selectLanguageName": "English",}}}中的selectLanguageName字段的值不对，在哪里改？

**Reply by MrHaoH** [reply]
大佬，这个问题我找到了，解决了，需要在plume.config.ts中手动设置selectLanguageName字段

**Reply by MrHaoH** [reply]
大佬，但现在又有个新问题，我看你源码中这样写的
import { useData, useSidebar } from '../../composables/index.js'

const { theme, site } = useData()
用theme, site来显示对应的多语言，我也用这种方式试了试，这是我的代码
import { useData } from 'vuepress-theme-plume/lib/client/composables/index.js'

const { theme, site } = useData()

但是报错：
17:59:01 [vite] Internal server error: Missing "./lib/client/composables/index.js" specifier in "vuepress-theme-plume" package
  Plugin: vite:import-analysis
  File: F:/lisinuo/pro-pinko/docs/.vuepress/components/nav-serve.vue
      at e (file:///F:/lisinuo/pro-pinko/node_modules/vite/dist/node/chunks/dep-DG6Lorbi.js:45978:25)
      at n (file:///F:/lisinuo/pro-pinko/node_modules/vite/dist/node/chunks/dep-DG6Lorbi.js:45978:627)
      at o (file:///F:/lisinuo/pro-pinko/node_modules/vite/dist/node/chunks/dep-DG6Lorbi.js:45978:1297)
      at resolveExportsOrImports (file:///F:/lisinuo/pro-pinko/node_modules/vite/dist/node/chunks/dep-DG6Lorbi.js:46599:18)
      at resolveDeepImport (file:///F:/lisinuo/pro-pinko/node_modules/vite/dist/node/chunks/dep-DG6Lorbi.js:46622:25)
      at tryNodeResolve (file:///F:/lisinuo/pro-pinko/node_modules/vite/dist/node/chunks/dep-DG6Lorbi.js:46387:16)
      at ResolveIdContext.resolveId (file:///F:/lisinuo/pro-pinko/node_modules/vite/dist/node/chunks/dep-DG6Lorbi.js:46137:19)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async PluginContainer.resolveId (file:///F:/lisinuo/pro-pinko/node_modules/vite/dist/node/chunks/dep-DG6Lorbi.js:48951:22)
      at async TransformPluginContext.resolve (file:///F:/lisinuo/pro-pinko/node_modules/vite/dist/node/chunks/dep-DG6Lorbi.js:49112:15)
跪求大佬回复

**Reply by pengzhanbo** [reply]
引入主题内部的 `composables` 的方式是：
```ts
import { useData } from 'vuepress-theme-plume/composables'
```
<https://theme-plume.vuejs.press/guide/api/client/#%E7%BB%84%E5%90%88%E5%BC%8F-api>

**Comment by sunshang-hl** [comment]
我在配置多语言时，提示以下错误。按照模板教程配置了config.ts中（'/fr/': { lang: 'fr-FR', title: 'Le blog' }, // 法语）部分，noters.ts中/* =================== locale: zh-TW ======================= */

const zhTWDemoNote = defineNoteConfig({
 .......})部分。以及各个readme.md文件下pageLayout: home配置中 link: /zh-TW/blog/的内容。请问时哪里出了问题？本页面中selectLanguageName的配置是必须的么？
vuepress-theme-plume:  ⚠ Locale /zh-TW/ is missing it's i18n config,
⠹ Initializing and preparing dataerror error in hook onPrepared from vuepress-theme-plume
✖ Initializing and preparing data - failed in 4.43s
error TypeError: Cannot read properties of undefined (reading 'home')

**Reply by sunshang-hl** [reply]
为何项目模板中没有selectLanguageName的配置，如果针对en和zh已内置，请在教程中予以说明。

**Reply by pengzhanbo** [reply]
默认适配的语言在 [多语言配置](https://theme-plume.vuejs.press/config/locales/#%E5%A4%9A%E8%AF%AD%E8%A8%80%E9%85%8D%E7%BD%AE) 有说明。另外这个错误信息跟 `selectLanguageName` 没有直接关系。这里是需要参照 [国际化配置](https://theme-plume.vuejs.press/guide/international/#%E4%B8%BB%E9%A2%98%E9%85%8D%E7%BD%AE) 中的 `locales`，其 key 是作为 页面访问路径前缀，同时也是该语言的存放文件夹。

**Reply by sunshang-hl** [reply]
反馈下：配置selectLanguageName后vuepress-theme-plume: ⚠ Locale /zh-TW/ is missing it's i18n config,
⠹ Initializing and preparing dataerror error in hook onPrepared from vuepress-theme-plume
✖ Initializing and preparing data - failed in 4.43s
error TypeError: Cannot read properties of undefined (reading 'home')的问题已经解决，项目能正常启动。
但是任然提示vuepress-theme-plume:  ⚠ Locale /zh-TW/ is missing it's i18n config， 而且最近项目启动会提示：@vuepress-plume/plugin-shikiji:  ⚠ You have not configured `plugins.shiki.languages`. It has been detected that you are using `["js","ts","html","css"]`. Please add it to the `plugins.shiki.languages` configuration. 以上错误提示反馈给你，可能需要修复下这个错误。

**Reply by pengzhanbo** [reply]
其它语言的默认适配我没有去做，虽然我也有另一个繁简体转换的开源项目，但是我比较熟悉的 中文繁体 和 台湾省的繁体 的区别很大，不单单只是 字的繁简体转换，词也需要改为另外的表达方式，而且还包含了不少的异体字，我对 zh-TW 并不熟，因此我没有对其内置支持。如果你熟练的话，可以向本仓库提 PR ，在 theme/src/node/locales/ 目录添加新的 多语言配置文件。

**Reply by pengzhanbo** [reply]
> @vuepress-plume/plugin-shikiji: ⚠ You have not configured plugins.shiki.languages. It has been detected that you are using ["js","ts","html","css"]. Please add it to the plugins.shiki.languages configuration. 

这个警告就是要告诉用户需要修改 `plugins.shiki.languages` 配置的，是主题主动抛出的警告信息，不是我需要修改，而是用户需要去修改。因为文档的代码块有使用到这些语言，所以需要去添加配置。而且生成的对应的配置文件我也有着重注释，目的就是要告诉用户需要着重注意这块的配置。
